Caboodle/Permissions/Permissions.android.cs

```diff
@@ -0,0 +1,165 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+using Android;
+using Android.Content.PM;
+using Android.Support.V4.App;
+using Android.Support.V4.Content;
+
+namespace Microsoft.Caboodle
+{
+    internal static partial class Permissions
+    {
+        static readonly object locker = new object();
+
+        static Dictionary<PermissionType, (int requestCode, TaskCompletionSource<PermissionStatus> tcs)> requests =
+            new Dictionary<PermissionType, (int, TaskCompletionSource<PermissionStatus>)>();
+
+        static void PlatformEnsureDeclared(PermissionType permission)
+        {
+            var androidPermissions = permission.ToAndroidPermissions(onlyRuntimePermissions: false);
+
+            // No actual android permissions required here, just return
+            if (androidPermissions == null || !androidPermissions.Any())
+                return;
+
+            var context = Platform.CurrentContext;
+
+            foreach (var ap in androidPermissions)
+            {
+                var packageInfo = context.PackageManager.GetPackageInfo(context.PackageName, PackageInfoFlags.Permissions);
+                var requestedPermissions = packageInfo?.RequestedPermissions;
+
+                // If the manifest is missing any of the permissions we need, throw
+                if (!requestedPermissions?.Any(r => r.Equals(ap, StringComparison.OrdinalIgnoreCase)) ?? false)
+                    throw new PermissionException($"You need to declare the permission: `{ap}` in your AndroidManifest.xml");
+            }
+        }
+
+        static Task<PermissionStatus> PlatformCheckStatusAsync(PermissionType permission)
+        {
+            PlatformEnsureDeclared(permission);
+
+            // If there are no android permissions for the given permission type
+            // just return granted since we have none to ask for
+            var androidPermissions = permission.ToAndroidPermissions(onlyRuntimePermissions: true);
+
+            if (androidPermissions == null || !androidPermissions.Any())
+                return Task.FromResult(PermissionStatus.Granted);
+
+            var context = Platform.CurrentContext;
+            var targetsMOrHigher = context.ApplicationInfo.TargetSdkVersion >= Android.OS.BuildVersionCodes.M;
+
+            foreach (var ap in androidPermissions)
+            {
+                if (targetsMOrHigher)
+                {
+                    if (context.CheckSelfPermission(ap) != Permission.Granted)
+                        return Task.FromResult(PermissionStatus.Denied);
+                }
+                else
+                {
+                    if (PermissionChecker.CheckSelfPermission(context, ap) != PermissionChecker.PermissionGranted)
```

Comment:
Is there a benefit to using the native method vs the support method? Won't the support fall back to the native method if it exists?

Reply:
@jamesmontemagno says there are dragons here and to do this.